src/core/net/dns_client.cpp

```diff
@@ -815,20 +863,69 @@ void Client::SendQuery(Query &aQuery, QueryInfo &aInfo, bool aUpdateTimer)
         SuccessOrExit(error = message->Append(Question(kQuestionRecordTypes[aInfo.mQueryType][num])));
     }
 
-    messageInfo.SetPeerAddr(aInfo.mConfig.GetServerSockAddr().GetAddress());
-    messageInfo.SetPeerPort(aInfo.mConfig.GetServerSockAddr().GetPort());
+    length = message->GetLength() - message->GetOffset();
+#if OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE
 
-    SuccessOrExit(error = mSocket.SendTo(*message, messageInfo));
+    // Send udp query in case of message being greater than kUdpQueryMaxSize or as last resort.
+    if (aInfo.mConfig.GetTransportProto() == QueryConfig::kDnsTcpTransport)
```

Comment:
I believe right now when user selects `kDnsTcpTransport` with DNS over TCP disabled it might be surprising as `Query` API will not throw the error but will use UDP instead.

Reply:
I have added alternative handling when this feature is turned off.